internal/service/alarms/internal/dictionary/dictionary.go

```
@@ -0,0 +1,190 @@
+				resultChannel <- result{
+					resourceTypeID: resource.id,
+					rules:          rules,
+					err:            err,
+				}
+			}()
+
+			switch resource.model {
+			case resourceTypeCluster:
+				rules, err = r.processCluster(ctx, resource.version)
+			default:
+				err = fmt.Errorf("unsupported resource type: %s", resource.model)
+			}
+		}(resource)
+	}
+
+	go func() {
+		wg.Wait()
+		close(resultChannel)
+	}()
+
+	for res := range resultChannel {
+		if res.err != nil {
+			slog.Error("error fetching rules for resource type", "ResourceType ID", res.resourceTypeID, "error", res.err)
+			continue
+		}
+
+		r.RulesMap[res.resourceTypeID] = res.rules
+		slog.Info("loaded rules for resource type", "ResourceType ID", res.resourceTypeID, "rules count", len(res.rules))
+	}
+
+	// TODO: Load future alarm dictionary struct
+	// TODO: Load alarm dictionary into DB
+}
+
+// processCluster processes a cluster resource type
+func (r *AlarmDictionary) processCluster(ctx context.Context, version string) ([]monitoringv1.Rule, error) {
+	cluster, err := r.getManagedCluster(ctx, version)
+	if err != nil {
+		return nil, err
+	}
+
+	promRules, err := r.getPromRules(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract rules from PrometheusRule list
+	var rules []monitoringv1.Rule
+	for _, promRule := range promRules {
+		for _, group := range promRule.Spec.Groups {
+			rules = append(rules, group.Rules...)
+		}
+	}
+
+	return rules, nil
+}
+
+// getManagedCluster finds a single managed cluster with the given version
+func (r *AlarmDictionary) getManagedCluster(ctx context.Context, version string) (*clusterv1.ManagedCluster, error) {
+	// Match managed cluster with the given version and not local cluster
+	selector := labels.NewSelector()
+	versionSelector, _ := labels.NewRequirement(managedClusterVersionLabel, selection.Equals, []string{version})
+	localClusterRequirement, _ := labels.NewRequirement(localClusterLabel, selection.NotEquals, []string{"true"})
```

can we return the errors?

---

The error happens when the rules for creating a label requirement constructor are violated. So, once you get this right, it won't fail. Unit tests are verifying this.